File: desktop/gatepath/diag/no_dns_probe.py
"""Reports a network that DHCP handed zero DNS servers.

Reads `ProbeContext.dns_server_count` and returns `NoDnsServers` when it is
zero — a half-broken connect where the captive redirect can never resolve,
so every later probe would be chasing a symptom rather than the cause.

Mirror of Android `NoDnsProbe.kt`.

Context-only — no network access.
"""

from __future__ import annotations

from gatepath.diag.probe import ProbeContext
from gatepath.diag.report import DiagnosticReport, Healthy, NoDnsServers


class NoDnsProbe:
    name = "no_dns"

    def run(self, ctx: ProbeContext) -> DiagnosticReport:
        if ctx.dns_server_count == 0:
            return NoDnsServers()
        return Healthy()

File: android/app/src/main/java/com/ventouxlabs/gatepath/diag/NoDnsProbe.kt
package com.ventouxlabs.gatepath.diag

/**
 * Reads [ProbeContext.dnsServerCount] and returns
 * [DiagnosticReport.NoDnsServers] when DHCP handed the network zero DNS
 * servers — a half-broken connect where the captive redirect can never
 * resolve.
 *
 * No network call — completes synchronously inside the per-probe budget.
 */
class NoDnsProbe : DiagnosticProbe {
    override val name = "no_dns"

    override suspend fun run(ctx: ProbeContext): DiagnosticReport =
        if (ctx.dnsServerCount == 0) {
            DiagnosticReport.NoDnsServers
        } else {
            DiagnosticReport.Healthy
        }
}
